pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -529,38 +530,51 @@ func (s *Controller) WorkloadInstanceHandler(wi *model.WorkloadInstance, event m
-				di.Endpoint.Address = addressToDelete
-				instancesDeleted = append(instancesDeleted, di)
-			}
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else if event == model.EventDelete {
-			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
-		} else {
-			s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, instance)
-		}
-		// If serviceentry's resolution is DNS, make a full push
-		// TODO: maybe cds?
-		if (se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN) &&
-			se.WorkloadSelector != nil {
-
-			fullPush = true
-			for _, inst := range instance {
-				configsUpdated[model.ConfigKey{
-					Kind:      kind.ServiceEntry,
-					Name:      string(inst.Service.Hostname),
-					Namespace: cfg.Namespace,
-				}] = struct{}{}
+		if se.WorkloadSelector != nil {
+			// If the workload selector exists, then we need to verify that the subset still is a match (in case the workload instance labels change)
+			if reflect.DeepEqual(se.WorkloadSelector.Labels, wi.Endpoint.Labels) || labels.Instance(se.WorkloadSelector.Labels).SubsetOf(wi.Endpoint.Labels) {
+				// If the workload instance still matches. We take care of the possible events.
+				seNamespacedName := config.NamespacedName(cfg)
+				services := s.services.getServices(seNamespacedName)
+				instance := convertWorkloadInstanceToServiceInstance(wi, services, se)
+				instances = append(instances, instance...)
+				if addressToDelete != "" {
+					for _, i := range instance {
+						di := i.DeepCopy()
+						di.Endpoint.Address = addressToDelete
+						instancesDeleted = append(instancesDeleted, di)
+					}
+					s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
+				} else if event == model.EventDelete {
+					s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
+				} else {
+					s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, instance)
+				}
+				// If serviceentry's resolution is DNS, make a full push
+				// TODO: maybe cds?
+				if (se.Resolution == networking.ServiceEntry_DNS || se.Resolution == networking.ServiceEntry_DNS_ROUND_ROBIN) &&
+					se.WorkloadSelector != nil {
+
+					fullPush = true
+					for _, inst := range instance {
+						configsUpdated[model.ConfigKey{
+							Kind:      kind.ServiceEntry,
+							Name:      string(inst.Service.Hostname),
+							Namespace: cfg.Namespace,
+						}] = struct{}{}
+					}
+				}
+			} else if oldWi != nil {
+				// If the labels don't match and there was an oldWi (not iterating the instance for the first time)
+				// then we see if we need to remove this instance
+				// Check if the old labels matched. If they did and the current labels don't. Then we need to remove this instance
+				if reflect.DeepEqual(se.WorkloadSelector.Labels, oldWi.Endpoint.Labels) || labels.Instance(se.WorkloadSelector.Labels).SubsetOf(oldWi.Endpoint.Labels) {
```

Comment:
Can you check se.WorkloadSelector.Labels.Equals(oldWi.Endpoint.Labels) instead of DeepEquals.

Reply:
Do you mean `labels.Instance(se.WorkloadSelector.Labels).Equals(oldWi.Endpoint.Labels)` ? (since se.WorkloadSelector.Labels is just a map and throws `type map[string]string has no field or method Equals`)

And `abels.Instance(se.WorkloadSelector.Labels).Equals()` actually calls `abels.Instance(se.WorkloadSelector.Labels).SubsetOf()` underneath the hood. So that brings us back to having a single `SubsetOf` check 😅